unikraft/app/application.go

```diff
@@ -634,3 +640,78 @@ func (app application) WithTarget(targ target.Target) (Application, error) {
 	ret.targets = target.Targets{targ.(*target.TargetConfig)}
 	return ret, nil
 }
+
+// MarshalYAML makes application implement yaml.Marshaller
+func (app application) MarshalYAML() (interface{}, error) {
+	ret := map[string]interface{}{
+		"specification": schema.SchemaVersionLatest,
+		"name":          app.name,
+		"unikraft":      app.unikraft,
+		"targets":       app.targets,
+		"libraries":     app.libraries,
+	}
+
+	// We purposefully do not marshal the configuration as this top level
+	// kconfig is redundant with the kconfig files in the libraries and
+	// unikraft.
+	// TODO: Figure out if anything is actually needed at this top level
+	// if app.configuration != nil && len(app.configuration) > 0 {
+	// 	ret["kconfig"] = app.configuration
+	// }
+
+	return ret, nil
+}
+
+func (app application) Save() error {
```

Comment:
```diff
 // MarshalYAML makes application implement yaml.Marshaller
 func (app application) MarshalYAML() (interface{}, error) {
-       ret := map[string]interface{}{
-               "specification": schema.SchemaVersionLatest,
-               "name":          app.name,
-               "unikraft":      app.unikraft,
-               "targets":       app.targets,
-               "libraries":     app.libraries,
+       unikraft, err := app.unikraft.MarshalYAML()
+       if err != nil {
+               return nil, err
+       }
+
+       var targets []interface{}
+       for _, targ := range app.targets {
+               t, err := targ.MarshalYAML()
+               if err != nil {
+                       return nil, err
+               }
+
+               targets = append(targets, t)
+       }
+
+       var libraries []interface{}
+       for _, lib := range app.libraries {
+               l, err := lib.MarshalYAML()
+               if err != nil {
+                       return nil, err
+               }
+
+               libraries = append(libraries, l)
        }
 
        // We purposefully do not marshal the configuration as this top level
@@ -659,7 +676,13 @@ func (app application) MarshalYAML() (interface{}, error) {
        //      ret["kconfig"] = app.configuration
        // }
 
-       return ret, nil
+       return map[string]interface{}{
+               "specification": schema.SchemaVersionLatest,
+               "name":          app.name,
+               "unikraft":      unikraft,
+               "targets":       targets,
+               "libraries":     libraries,
+       }, nil
 }
 
 func (app application) Save() error {
 ```

Reply:
Hi @nderjung! Thank you for the review. I added the ``yamlMarshaler`` to the ``Component`` interface. As for the error checking during the marshalling, both from my testing and the [documentation](https://pkg.go.dev/gopkg.in/yaml.v2#Marshaler), this is already the default behaviour: 
> If an error is returned by MarshalYAML, the marshaling procedure stops and returns with the provided error. 

Should I still add the extra checks?

I also refactored the top level function to avoid marshalling empty arrays.